Bento/Diff/CollectionViewSectionDiff.swift

```diff
@@ -0,0 +1,108 @@
+import UIKit
+import FlexibleDiff
+
+struct CollectionViewSectionDiff<SectionId: Hashable, RowId: Hashable> {
+    private let oldSections: [Section<SectionId, RowId>]
+    private let newSections: [Section<SectionId, RowId>]
+
+    init(oldSections: [Section<SectionId, RowId>],
+         newSections: [Section<SectionId, RowId>]) {
+        self.oldSections = oldSections
+        self.newSections = newSections
+    }
+
+    func apply(to collectionView: UICollectionView, completion: (() -> Void)? = nil) {
+        let diff = SectionedChangeset(previous: oldSections,
+                                      current: newSections,
+                                      sectionIdentifier: { $0.id },
+                                      areMetadataEqual: Section.hasEqualMetadata,
+                                      items: { $0.rows },
+                                      itemIdentifier: { $0.id },
+                                      areItemsEqual: ==)
+        apply(diff: diff, to: collectionView, completion: completion)
+    }
+
+    private func apply(diff: SectionedChangeset, to collectionView: UICollectionView, completion: (() -> Void)?) {
+        collectionView.performBatchUpdates({
+            for (item, section) in diff.sections.mutations.enumerated() {
+                if let headerView = collectionView.supplementaryView(forElementKind: UICollectionElementKindSectionHeader.description,
+                                                                     at: IndexPath(row: item, section: section)),
+                    let node = newSections[section].header {
+                    update(view: headerView, with: node)
+                }
+                if let footerView = collectionView.supplementaryView(forElementKind: UICollectionElementKindSectionFooter.description,
+                                                                     at: IndexPath(row: item, section: section)),
+                    let node = newSections[section].footer {
+                    update(view: footerView, with: node)
+                }
+            }
+
+            collectionView.insertSections(diff.sections.inserts)
+            collectionView.deleteSections(diff.sections.removals)
+            apply(sectionMutations: diff.mutatedSections, to: collectionView)
+            collectionView.moveSections(diff.sections.moves)
+
+        }, completion: { _ in completion?() })
+    }
+
+    private func apply(sectionMutations: [SectionedChangeset.MutatedSection],
+                       to collectionView: UICollectionView) {
+        for sectionMutation in sectionMutations {
+
+            collectionView.deleteItems(at: sectionMutation.deletedIndexPaths)
+            collectionView.insertItems(at: sectionMutation.insertedIndexPaths)
+            collectionView.perform(moves: sectionMutation.movedCollectionIndexPaths)
+            [sectionMutation.changeset.moves.lazy
```

Comment:
I found this bit of code really hard to read. Maybe create an intermediate variable `sectionChanges` with everything up to `joined` and then call `sectionChanges.forEach {}` ?